### PR TITLE
Resolve requirement names from distribution metadata

### LINKED ISSUE #141: Module name are not required to match import name
For requirements.txt generation we should consider https://x.com/charliermarsh/status/1934456910566039821?s=46&t=PWw_cbXuOb-YVHQ7BKTG7g

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

- separate requirement-name inference from module classification
- resolve imports through the selected Python environment using Core Metadata 2.5 `Import-Name` / `Import-Namespace`, installed-file records, and legacy `top_level.txt`
- add `--python`, `CRIBO_PYTHON`, and `[requirements]` configuration with longest-prefix `module-map` overrides
- reject equally strong providers with an actionable error instead of choosing by scan order
- add end-to-end fixtures for mismatched distribution names and ambiguous namespace providers

## Why

Requirements generation previously assumed that a top-level import name was also the installable distribution name. That produces incorrect output for packages such as `sklearn` / `scikit-learn`, and filesystem scan order cannot safely resolve namespaces supplied by multiple distributions.

The new resolver uses metadata from the environment that will run the bundle, while retaining explicit configuration for missing or ambiguous metadata.

Closes #141

## User impact

`--emit-requirements` now emits normalized distribution names when installed metadata identifies them. Users can select an interpreter explicitly with `--python`, `CRIBO_PYTHON`, or `requirements.python`; `python -m cribo` automatically supplies its own interpreter. Ambiguous providers fail with guidance to configure `requirements.module-map`, while unmapped imports retain a valid normalized fallback.

## Validation

- `cargo nextest run --workspace` (194 passed)
- `cargo insta test --all-features --check --unreferenced reject` (370 passed; no unreferenced snapshots)
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- `cargo fmt --all -- --check`
- `uvx ruff check python/cribo/__main__.py crates/cribo/src/requirement_resolver.py`

## Summary by CodeRabbit

* **New Features**
  * Added automatic generation of Python requirements from imported modules.
  * Added `--python` and `CRIBO_PYTHON` options for selecting the Python interpreter.
  * Added configurable import-to-requirement mappings with longest-prefix matching.
  * Improved detection across virtual environments and installed package metadata.
  * Added clear handling for ambiguous or invalid package mappings.

* **Documentation**
  * Documented the new command-line option, configuration settings, and environment variable.

* **Bug Fixes**
  * Improved package ownership detection and requirements selection across multiple environments.

### CODE REVIEW

**Comment by tinovyatkin** [comment]
@codex review

File: crates/cribo/src/requirement_resolver.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    for path in request["metadata_paths"] + sys.path:
+        if path and path not in search_paths:
+            search_paths.append(path)
+
+    distributions = list(importlib.metadata.distributions(path=search_paths))
+    resolutions = {
+        import_name: distribution_candidates(import_name, distributions)
+        for import_name in sorted(request["imports"])
+    }
+    json.dump({"resolutions": resolutions}, sys.stdout, sort_keys=True)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py]: Addressed in `d6f5d2a`. The helper now reads and writes JSON through binary stdio with explicit UTF-8 encoding. The non-ASCII protocol regression passes; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    depth = import_name.count(".") + 1
+    if path == prefix + ".py" or path == prefix + "/__init__.py":
+        return 4000 + depth, "installed file"
+
+    for suffix in importlib.machinery.EXTENSION_SUFFIXES:
+        if path == prefix + suffix or path == prefix + "/__init__" + suffix:
+            return 4000 + depth, "installed extension"
+
+    if path.startswith(prefix + "/"):
+        return 3000 + depth, "installed namespace descendant"
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py]: Addressed in `d6f5d2a`. Absolute package paths are relativized against normalized import roots before root-anchored matching. Coverage verifies an editable absolute path and rejects nested paths such as `pandera/api/polars`; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                    candidates,
+                    project_name,
+                    2000 + depth,
+                    "core metadata Import-Namespace",
+                )
+
+        try:
+            files = distribution.files or ()
+        except (OSError, ValueError):
+            files = ()
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py:170]: Addressed in `d6f5d2a`. Access to `distribution.files` now catches general metadata exceptions and continues with no file evidence. The helper regression raises `RuntimeError` here; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        for package_path in files:
+            evidence = file_score(import_name, normalized_path(package_path))
+            if evidence is not None:
+                score, description = evidence
+                add_candidate(candidates, project_name, score, description)
+
+        try:
+            top_level = distribution.read_text("top_level.txt") or ""
+        except (OSError, ValueError):
+            top_level = ""
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py:197]: Addressed in `d6f5d2a`. Reading `top_level.txt` now catches general metadata exceptions and continues without legacy evidence. The helper regression raises `RuntimeError` here; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/resolver.rs [hunk 5]
```diff
@@ -1355,52 +1285,54 @@ impl ModuleResolver {
...
+                .and_then(|value| value.split(';').next())
+                .is_some_and(|prefix| {
+                    !prefix.is_empty()
+                        && (import_name == prefix
+                            || import_name
+                                .strip_prefix(prefix)
+                                .is_some_and(|suffix| suffix.starts_with('.')))
+                })
+        })
     }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/resolver.rs]: Addressed in `d6f5d2a`. Metadata lines are split at the colon and `Import-Name` / `Import-Namespace` are matched with `eq_ignore_ascii_case`, without allocating lowercase copies. Lowercase and uppercase header regressions pass; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.rs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+            }
+        }
+
+        if let Some(first_path) = self.metadata_paths.first() {
+            for ancestor in first_path.ancestors() {
+                let candidate = Self::environment_python(&ancestor.join(".venv"));
+                if candidate.is_file() {
+                    return Ok(candidate);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.rs]: Addressed in `d6f5d2a`. Requirement resolution now shares the resolver virtualenv-name list and probes `.venv`, `venv`, `env`, `.virtualenv`, and `virtualenv` before PATH fallback. A regression exercises every name; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/resolver.rs [hunk 7]
```diff
@@ -969,11 +961,10 @@ impl ModuleResolver {
     ) -> ImportClassification {
         let explicit_first_party = self.config.known_first_party.contains(module_name);
         let explicit_third_party = self.config.known_third_party.contains(module_name);
-        let root_import = module_name.split('.').next().unwrap_or(module_name);
-        let distribution = self.find_package_name_in_site_packages(search_root, root_import);
+        let has_distribution = self.distribution_owns_import(search_root, module_name);
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/resolver.rs]: Addressed in `d6f5d2a`. The origin condition now short-circuits before `distribution_owns_import` when the module is explicitly third-party or the search location already has third-party origin. Full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/config.rs [hunk 8]
```diff
@@ -159,6 +183,10 @@ impl EnvConfig {
             config.tree_shake = parse_bool(&tree_shake_str);
         }
 
+        if let Ok(python) = env::var("CRIBO_PYTHON") {
+            config.python = Some(PathBuf::from(python));
+        }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/config.rs:206]: Addressed in `d6f5d2a`. Empty `CRIBO_PYTHON` values now parse as no override, while non-empty paths remain unchanged. The focused environment-path regression passes; full validation: nextest 199/199, Insta 380/380, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.rs [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+
+        Err(anyhow!(
+            "Could not find a Python interpreter for distribution metadata; use --python or set \
+             requirements.python"
+        ))
+    }
+
+    fn auto_detected_python(&self) -> Option<PathBuf> {
+        let first_path = self.metadata_paths.first()?;
+        for ancestor in first_path.ancestors() {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.rs:307]: Addressed in `119ec49`. Requirement resolution now receives the resolver-selected virtualenv site-packages alongside normal search roots. A CLI regression launches from a directory containing an unactivated `.venv` while bundling an external entry and verifies the distinct distribution name. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/config.rs [hunk 10]
```diff
@@ -38,6 +38,20 @@ pub struct Config {
...
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(default)]
+pub struct RequirementsConfig {
+    /// Python interpreter whose environment supplies distribution metadata
+    pub python: Option<PathBuf>,
+
+    /// Explicit import-prefix to PEP 508 requirement mappings
+    #[serde(rename = "module-map")]
+    pub module_map: IndexMap<String, String>,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/config.rs]: Addressed in `119ec49`. `RequirementsConfig` now uses a custom `Debug` implementation that preserves the interpreter path and only reports a redacted module-map entry count. The regression verifies that neither mapping keys nor a token-bearing URL appear in `Config` debug output. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        },
+        key=len,
+        reverse=True,
+    )
+
+    distributions = list(importlib.metadata.distributions(path=search_paths))
+    resolutions = {
+        import_name: distribution_candidates(import_name, distributions, import_roots)
+        for import_name in sorted(request["imports"])
+    }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py]: Addressed in `119ec49`. The Python helper now builds one reusable ownership index from headers, importable files, and `top_level.txt` before resolving imports; the Rust resolver caches one ownership index per canonical search root. Regressions verify each Python metadata source is read once across multiple imports and that Rust lookups reuse the cached index. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            import_prefix = declared_name(value)
+            if import_prefix and is_import_prefix(import_prefix, import_name):
+                depth = import_prefix.count(".") + 1
+                add_candidate(
+                    candidates,
+                    project_name,
+                    2000 + depth,
+                    "core metadata Import-Namespace",
+                )
+
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py:166]: Addressed in `119ec49`. Access to `distribution.metadata` and its header methods is now guarded; an unreadable distribution is skipped without aborting resolution. The helper regression exercises a metadata property that raises `RuntimeError`. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py]: Addressed in `119ec49`. The metadata index now detects names declared in both `Import-Name` and `Import-Namespace` and fails with a concise distribution-specific error before scoring candidates. An end-to-end metadata-query regression covers the conflict. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.rs [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+        None
+    }
+
+    fn environment_python(environment: &Path) -> PathBuf {
+        if cfg!(windows) {
+            environment.join("Scripts").join("python.exe")
+        } else {
+            environment.join("bin").join("python")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.rs:325]: Addressed in `119ec49`. Added doc comments for every introduced `RequirementResolver` method, environment-path parsing, import-classification construction and bundle policy, and the new distribution metadata indexing/cache helpers. Full validation: nextest 203/203, Insta 387/387, strict Clippy, fmt, and Ruff.

File: python/cribo/__main__.py [hunk 14]
```diff
@@ -5,15 +5,18 @@
...
 
 
 def main() -> None:
     """Main entry point that delegates to the cribo binary."""
     try:
+        env = os.environ.copy()
+        env.setdefault("CRIBO_PYTHON", sys.executable)
         # Call the cribo binary with the same arguments
-        result = subprocess.run(["cribo"] + sys.argv[1:], check=False)
+        result = subprocess.run(["cribo"] + sys.argv[1:], check=False, env=env)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply python/cribo/__main__.py:20]: Addressed in `119ec49`. The Python module wrapper now substitutes `sys.executable` when `CRIBO_PYTHON` is unset or empty and preserves any non-empty override. Both cases were exercised directly, and Ruff passes. Full repository validation: nextest 203/203, Insta 387/387, strict Clippy, and fmt.

File: crates/cribo/src/requirement_resolver.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+    import_roots = sorted(
+        {
+            normalized_path(os.path.abspath(path)).rstrip("/")
+            for path in search_paths
+        },
+        key=len,
+        reverse=True,
+    )
+
+    distributions = list(importlib.metadata.distributions(path=search_paths))
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py]: Addressed in `79163ee`. Requirement generation now carries the concrete `ModuleResolver` search root into per-root metadata indexes and prefers candidates from that root, while retaining all-root fallback for missing/editable metadata and ambiguity for namespace packages. A CLI regression covers two `PYTHONPATH` roots that both claim the same import and verifies the first root’s distribution wins. Full validation: workspace nextest, Insta 388/388 with no changed or unreferenced snapshots, strict Clippy, fmt, and Ruff.

File: crates/cribo/src/requirement_resolver.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+            }
+        except Exception:
+            continue
+        if not project_name:
+            continue
+
+        conflicting_names = import_names & namespace_names
+        if conflicting_names:
+            names = ", ".join(sorted(conflicting_names))
+            raise InvalidDistributionMetadata(
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/requirement_resolver.py:140]: Addressed in `f7b5b67`. Metadata conflicts are now checked only against prefixes of requested imports; unrelated conflicting declarations are excluded from scoring without aborting requirement generation. Added an end-to-end metadata-query regression. Full validation: workspace nextest 205/205, Insta 390/390 with no changed or unreferenced snapshots, strict Clippy, fmt, and Ruff.
